### PR TITLE
SILOptimizer: restructure the apply(partial_apply) peephole and the dead partial_apply elimination optimizations

### PR DESCRIPTION
**Pull request by eeckstein**
Changes:

* Allow optimizing partial_apply capturing opened existential: we didn't do this originally because it was complicated to insert the required alloc/dealloc_stack instructions at the right places. Now we have the StackNesting utility, which makes this easier.

* Support indirect-in parameters. Not super important, but why not? It's also easy to do with the StackNesting utility.

* Share code between dead closure elimination and the apply(partial_apply) optimization. It's a bit of refactoring and allowed to eliminate some code which is not used anymore.

* Fix an ownership problem: We inserted copies of partial_apply arguments _after_ the partial_apply (which consumes the arguments).

* When replacing an apply(partial_apply) -> apply and the partial_apply becomes dead, avoid inserting copies of the arguments twice.

These changes don't have any immediate effect on our current benchmarks, but will allow eliminating curry thunks for existentials.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

File: include/swift/SILOptimizer/Utils/InstOptUtils.h [hunk 1]
```diff
@@ -295,16 +295,36 @@ struct InstModCallbacks {
...
+                                 bool includeTrivialAddrArgs);
+
+/// Collect all (transitive) users of \p inst which just copy or destroy \p inst.
+///
+/// In other words: all users which do not prevent \p inst from being considered
+/// as "dead".
+/// Returns true, if there are no other users beside those collected in \p
+/// destroys, i.e. if \p inst can be considered as "dead".
+bool collectDestroys(SingleValueInstruction *inst,
+                             SmallVectorImpl<SILInstruction *> &destroys);
```
Comment (gottesmm) [inline include/swift/SILOptimizer/Utils/InstOptUtils.h]: Can you run this patch through git-clang-format?

**Review by gottesmm — COMMENTED** [review]
Quick nit. I need to read this offline in my editor. I'll do a more in depth in a bit.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

File: include/swift/SILOptimizer/Utils/InstOptUtils.h (hunk 1 above)
Comment (eeckstein) [reply include/swift/SILOptimizer/Utils/InstOptUtils.h]: okay, done

File: lib/SILOptimizer/Utils/ValueLifetime.cpp [hunk 2]
```diff
@@ -281,3 +289,20 @@ void ValueLifetimeAnalysis::dump() const {
   }
   llvm::errs() << '\n';
 }
+
+void swift::endLifetimeAtFrontier(
+    SILValue valueOrStackLoc, const ValueLifetimeAnalysis::Frontier &frontier,
+    SILBuilder &builder) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/ValueLifetime.cpp]: Why pass in a SILBuilder here? Better to pass in a SILBuilderContext and use a locally created SILBuilderWithScope. Also, why are you messing with RegularLocations here?

File: lib/SILOptimizer/Utils/ValueLifetime.cpp [hunk 3]
```diff
@@ -190,7 +197,8 @@ bool ValueLifetimeAnalysis::computeFrontier(Frontier &frontier, Mode mode,
 
     for (unsigned i = 0, e = succBlocks.size(); i != e; ++i) {
       if (unhandledFrontierBlocks.count(succBlocks[i])) {
-        assert(isCriticalEdge(term, i) && "actually not a critical edge?");
+        assert((isCriticalEdge(term, i) || userSet.count(term)) &&
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/ValueLifetime.cpp]: Can you explain this to me? Why is this change needed? Does userSet.count(term) mean a critical edge?

File: lib/SILOptimizer/Utils/ValueLifetime.cpp [hunk 4]
```diff
@@ -135,8 +135,15 @@ bool ValueLifetimeAnalysis::computeFrontier(Frontier &frontier, Mode mode,
         continue;
       }
       // In case the last user is a TermInst we add all successor blocks to the
-      // frontier (see below).
-      assert(deadInSucc && "The final using TermInst must have successors");
+      // frontier (see below). Except it is a function exiting TermInst, like
+      // 'return' or 'throw'.
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/ValueLifetime.cpp]: Can you expand the comment here why the term inst is an exception. I understand why, it should just be explicit in the code.

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp [hunk 5]
```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
...
-  for (auto arg : pai->getArguments()) {
-    // If there is new temporary for this argument, use it instead.
-    if (argToTmpCopy.count(arg)) {
-      arg = argToTmpCopy.lookup(arg);
-    }
-    argList.push_back(arg);
-  }
-
-  builder.setInsertionPoint(insertPoint);
+  builder.setInsertionPoint(paiAI.getInstruction());
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: Please change this to use a new SILBuilderWithScope that takes in a SILBuilderContext.

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp [hunk 6]
```diff
@@ -229,78 +139,80 @@ bool PartialApplyCombiner::processSingleApply(FullApplySite paiAI) {
...
-                              cast<ApplyInst>(paiAI)->isNonThrowing());
-
+  if (auto *tai = dyn_cast<TryApplyInst>(paiAI)) {
+    builder.createTryApply(paiAI.getLoc(), callee, subs, argList,
+                           tai->getNormalBB(), tai->getErrorBB());
+  } else {
+    auto *apply = cast<ApplyInst>(paiAI);
+    auto *newAI = builder.createApply(paiAI.getLoc(), callee, subs, argList,
+                                      apply->isNonThrowing());
+    callbacks.replaceValueUsesWith(apply, newAI);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: Did you add this callback due to a test failure? Which test was it?

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp [hunk 7]
```diff
@@ -347,25 +259,29 @@ SILInstruction *PartialApplyCombiner::combine() {
...
-    if (!processSingleApply(ai))
-      return nullptr;
+    foundApplySites.push_back(ai);
   }
 
-  // release/destroy and deallocate introduced temporaries.
-  if (!tmpCopies.empty()) {
-    destroyTemporaries();
-    deallocateTemporaries();
+  if (foundApplySites.empty())
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: I like the way this loop is done here. Much better!

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 8]
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
...
+    SILParameterInfo param = paramList[i];
+    if (param.isIndirectMutating())
       continue;
 
-    // Now that we know that we have a non-argument address, perform a take-init
-    // of arg into a lifetime extended alloc_stack
-    args[i] = createLifetimeExtendedAllocStack(builder, loc, arg, exitingBlocks,
-                                               callbacks);
+    SILType ty = argOp.get()->getType();
+    if (!ty.isTrivial(*F) || (includeTrivialAddrArgs && ty.isAddress()))
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: This is really nice here!

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 9]
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
...
-      releasePartialApplyCapturedArg(builder, loc, arg, param, callbacks);
+bool swift::collectDestroys(SingleValueInstruction *inst,
+                            SmallVectorImpl<SILInstruction *> &destroys) {
+  bool isDead = true;
+  for (Operand *use : inst->getUses()) {
+    SILInstruction *user = use->getUser();
+    if (useHasTransitiveOwnership(user)) {
+      if (!collectDestroys(cast<SingleValueInstruction>(user), destroys))
+        isDead = false;
+      destroys.push_back(user);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: Rather than else if, can you use early continues and a worklist instead of recursion?

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 10]
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
...
     return false;
-  deleteInsts.push_back(inst);
-  for (auto *use : cast<SingleValueInstruction>(inst)->getUses()) {
-    if (!deadMarkDependenceUser(use->getUser(), deleteInsts))
-      return false;
+  }
+
+  for (Operand *argOp : argsToHandle) {
+    SILValue arg = argOp->get();
+    int argIdx = argOp->getOperandNumber() - pai->getArgumentOperandNumber();
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: Please do not hard code this directly. This should definitely be something on function convention or apply site apply or something like that. @atrick your thoughts?

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 11]
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
...
+    SILDebugVariable dbgVar(/*Constant*/ true, argIdx);
+    builder.setInsertionPoint(pai);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: We are trying to move away from code like this. Can you use a new SILBuilderWithScope if possible?

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 12]
```diff
@@ -1521,38 +1510,53 @@ bool swift::tryDeleteDeadClosure(SingleValueInstruction *closure,
...
   // If we have a partial_apply, release each captured argument at each one of
   // the final release locations of the partial apply.
   if (auto *pai = dyn_cast<PartialApplyInst>(closure)) {
-    // If we can not decrement the ref counts of the dead partial apply for any
-    // reason, bail.
-    if (!releaseCapturedArgsOfDeadPartialApply(pai, tracker, callbacks))
-      return false;
+    assert(!pa->isOnStack() &&
+           "partial_apply [stack] should have been handled before");
+    SILBuilder builder(pai);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: SILBuilder => SILBuilderWithScope

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 13]
```diff
@@ -1521,38 +1510,53 @@ bool swift::tryDeleteDeadClosure(SingleValueInstruction *closure,
...
+    if (needKeepArgsAlive) {
+      if (!keepArgsOfPartialApplyAlive(pai, closureDestroys, builder))
+        return false;
+    } else {
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: No else after return according to LLVM style guide. Please remove this.

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 14]
```diff
@@ -1521,38 +1510,53 @@ bool swift::tryDeleteDeadClosure(SingleValueInstruction *closure,
...
+      SmallVector<Operand *, 8> argsToHandle;
+      getConsumedPartialApplyArgs(pai, argsToHandle,
+                                  /*includeTrivialAddrArgs*/ false);
+
+      // We can just destroy the arguments at the point of the partial_apply
+      // (remember: partial_apply consumes all arguments).
+      for (Operand *argOp : argsToHandle) {
+        SILValue arg = argOp->get();
+        if (arg->getType().isObject()) {
+          builder.emitDestroyValueOperation(pai->getLoc(), arg);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: I think we do this a lot. Add this as a helper on SILBuilder? Maybe emitDestroyOperation?

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp [hunk 15]
```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
...
 
-    for (auto copy : tmpCopies) {
-      builder.setInsertionPoint(term);
-      builder.createDeallocStack(pai->getLoc(), copy);
+    SILValue tmp = arg;
+    if (arg->getType().isObject()) {
+      tmp = builder.emitCopyValueOperation(pai->getLoc(), arg);
+    } else {
+      // Copy address-arguments into a stack-allocated temporary.
+      tmp = builder.createAllocStack(pai->getLoc(), arg->getType(), dbgVar);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: I am fine with this now, but I do think that at some point, we should incorporate code from SILGen in terms of ManagedValue/Scopes here. Would make it much easier to work with.

File: include/swift/SILOptimizer/Utils/ValueLifetime.h [hunk 16]
```diff
@@ -79,6 +79,9 @@ class ValueLifetimeAnalysis {
   /// instructions of the frontier that are not in the critical edges. Note that
   /// the method getCriticalEdges can be used to retrieve the critical edges.
   ///
+  /// An edge is also considered as "critical" if it has a single precedessor
+  /// but the predecessor's terminal instruction is a user the value.
```
Comment (gottesmm) [inline include/swift/SILOptimizer/Utils/ValueLifetime.h]: Can you elaborate on this here in the comment. The comment is defining something without explaining why. An example would be useful.

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp [hunk 17]
```diff
@@ -229,78 +139,80 @@ bool PartialApplyCombiner::processSingleApply(FullApplySite paiAI) {
...
       // Copy the argument as the callee may consume it.
-      arg = builder.emitCopyValueOperation(pai->getLoc(), arg);
-      // For non consumed parameters (e.g. guaranteed), we also need to
-      // insert destroys after each apply instruction that we create.
-      if (!paramInfo[paramInfo.size() - partialApplyArgs.size() + i]
-               .isConsumed())
-        toBeDestroyedArgs.push_back(arg);
+      if (arg->getType().isAddress()) {
+        auto *ASI = builder.createAllocStack(pai->getLoc(), arg->getType());
+        builder.createCopyAddr(pai->getLoc(), arg, ASI, IsTake_t::IsNotTake,
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: Thanks for fixing this. We should be copying before each call site. This would be another place where SILGen's ManagedValue/Scope machinery would be useful.

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp [hunk 18]
```diff
@@ -229,78 +139,80 @@ bool PartialApplyCombiner::processSingleApply(FullApplySite paiAI) {
...
 
-  FullApplySite nai;
-  if (auto *tai = dyn_cast<TryApplyInst>(paiAI))
-    nai = builder.createTryApply(paiAI.getLoc(), callee, subs, argList,
-                                 tai->getNormalBB(), tai->getErrorBB());
-  else
-    nai = builder.createApply(paiAI.getLoc(), callee, subs, argList,
-                              cast<ApplyInst>(paiAI)->isNonThrowing());
-
+  if (auto *tai = dyn_cast<TryApplyInst>(paiAI)) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: Can you change this to be a full switch over the type of full apply site with llvm_unreachable in the cases which we do not support today. That way people know to update this code in the future.

File: lib/SILOptimizer/Mandatory/MandatoryInlining.cpp [hunk 19]
```diff
@@ -342,7 +342,8 @@ static SILValue cleanupLoadedCalleeValue(SILValue calleeValue, LoadInst *li) {
 
 /// Removes instructions that create the callee value if they are no
 /// longer necessary after inlining.
-static void cleanupCalleeValue(SILValue calleeValue) {
+static void cleanupCalleeValue(SILValue calleeValue,
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/MandatoryInlining.cpp]: Why not return needUpdateStackNesting as a bool?

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp [hunk 20]
```diff
@@ -104,10 +104,18 @@ SILInstruction *SILCombiner::visitPartialApplyInst(PartialApplyInst *PAI) {
...
+      tryOptimizeApplyOfPartialApply(PAI, Builder, getInstModCallbacks());
+  if (argsAreKeptAlive)
+    needUpdateStackNesting = true;
+
+  // Try to delete the partial_apply.
+  // In case it became dead because of tryOptimizeApplyOfPartialApply, we don't
+  // need to copy all arguments again (to extend their lifetimes), because it
+  // was already done in tryOptimizeApplyOfPartialApply.
+  if (tryDeleteDeadClosure(PAI, getInstModCallbacks(), !argsAreKeptAlive))
+    needUpdateStackNesting = true;
```
Comment (gottesmm) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: needUpdateStackNesting => stackNestingInvalidated?

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 21]
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
...
+  SmallVector<Operand *, 8> argsToHandle;
+  getConsumedPartialApplyArgs(pai, argsToHandle,
+                              /*includeTrivialAddrArgs*/ false);
+  if (argsToHandle.empty())
+    return true;
+
+  // Compute the set of endpoints, which will be used to insert destroys of
+  // temporaries. This may fail if the frontier is located on a critical edge
+  // which we may not split.
+  ValueLifetimeAnalysis vla(pai, paiUsers);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: Thank you for fixing this. I was worried about this code using ValueLifetimeAnalysis(SILValue) API. We should really get rid of that entry point since it is so fragile and easy to mess up.

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 22]
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
...
-/// For each captured argument of pai, decrement the ref count of the captured
-/// argument as appropriate at each of the post dominated release locations
-/// found by tracker.
-static bool releaseCapturedArgsOfDeadPartialApply(PartialApplyInst *pai,
-                                                  ReleaseTracker &tracker,
-                                                  InstModCallbacks callbacks) {
-  SILBuilderWithScope builder(pai);
-  SILLocation loc = pai->getLoc();
+static bool
+deadMarkDependenceUser(SILInstruction *inst,
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: Is this dead now? I noticed you got rid of the usage below?

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 23]
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
...
+/// Move the original arguments of the partial_apply into newly created
+/// temporaries to extend the lifetime of the arguments until the partial_apply
+/// is finally destroyed.
+///
+/// TODO: figure out why this is needed at all. Probably because of some
+///       weirdness of the old retain/release ARC model. Most likely this will
+///       not be needed anymore with OSSA.
+static bool keepArgsOfPartialApplyAlive(PartialApplyInst *pai,
+                                        ArrayRef<SILInstruction *> paiUsers,
+                                        SILBuilder &builder) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: Another case where I would suggest passing in a SILBuilderContext and creating builders locally. Will prevent Debug Info bugs.

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp [hunk 24]
```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
...
-    if (arg->getType().hasOpenedExistential())
-      return false;
-
-    // If the temporary is non-trivial, we need to destroy it later.
-    if (!arg->getType().isTrivial(*pai->getFunction()))
-      needsDestroys = true;
-    argsToHandle.push_back(std::make_pair(arg, i));
+  // Of course we must inlude all apply instructions which we want to optimize.
+  for (FullApplySite ai : applies) {
+    paiUsers.push_back(ai.getInstruction());
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: What about this:

```
transform(applies, std::back_inserter(paiUsers), [](FullApplySite ai) { return ai.getInstruction(); })
```

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp [hunk 25]
```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
...
-bool PartialApplyCombiner::allocateTemporaries() {
+/// Copy the original arguments of the partial_apply into newly created
+/// temporaries and use these temporaries instead of the original arguments
+/// afterwards.
+///
+/// This is done to "extend" the life-time of original partial_apply arguments,
+/// as they may be destroyed/deallocated before the last use by one of the
+/// apply instructions.
+bool PartialApplyCombiner::copyArgsToTemporaries(
+    const SmallVectorImpl<FullApplySite> &applies) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: => ArrayRef<FullApplySite>

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 26]
```diff
@@ -1407,92 +1377,111 @@ void swift::releasePartialApplyCapturedArg(SILBuilder &builder, SILLocation loc,
...
-    SILValue arg = args[i];
-    SILParameterInfo paramInfo = params[i];
-
-    // If we are not going to destroy this partial_apply, continue.
-    if (!shouldDestroyPartialApplyCapturedArg(arg, paramInfo,
-                                              builder.getFunction()))
-      continue;
+  ArrayRef<SILParameterInfo> paramList = paiTy->getParameters();
+  auto argList = pai->getArgumentOperands();
+  paramList = paramList.drop_front(paramList.size() - argList.size());
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp]: Can we make an API around this on ApplySite? Or on partial apply?

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp [hunk 27]
```diff
@@ -35,192 +32,105 @@ class PartialApplyCombiner {
...
-/// Emit dealloc_stack for all temporaries.
-void PartialApplyCombiner::deallocateTemporaries() {
-  // Insert dealloc_stack instructions at all function exit points.
-  for (SILBasicBlock &block : *pai->getFunction()) {
-    TermInst *term = block.getTerminator();
-    if (!term->isFunctionExiting())
-      continue;
+  for (Operand *argOp : argsToHandle) {
+    SILValue arg = argOp->get();
+    int argIdx = argOp->getOperandNumber() - pai->getArgumentOperandNumber();
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: Please don't hard code this. This should be sort sort of higher level API on function conventions or ApplySite.

File: lib/SILOptimizer/Utils/ValueLifetime.cpp (hunk 2 above)
Comment (eeckstein) [reply lib/SILOptimizer/Utils/ValueLifetime.cpp]: SILBuilder: ok, makes sense.

The frontier instruction can have a ReturnLocation. Therefore I have to construct a RegularLocation.

File: lib/SILOptimizer/Utils/ValueLifetime.cpp (hunk 3 above)
Comment (eeckstein) [reply lib/SILOptimizer/Utils/ValueLifetime.cpp]: This is a bug fix.
```
/// An edge is also considered as "critical" if it has a single precedessor
/// but the predecessor's terminal instruction is a user of the value.
///
```

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp (hunk 24 above)
Comment (eeckstein) [reply lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: I find the for-loop version easier to understand.

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp (hunk 6 above)
Comment (eeckstein) [reply lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: This callback was already in the original code.

**Comment by eeckstein** [comment]
@gottesmm Thanks for the review. I have addressed your points in this new version. Much of the code is just copied from the original version, which is already quite old. That's why SILBuilderContext etc. was not used. But, yes, it makes sense to refresh the code.

**Comment by eeckstein** [comment]
@swift-ci smoke test

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp [hunk 28]
```diff
@@ -104,10 +104,20 @@ SILInstruction *SILCombiner::visitPartialApplyInst(PartialApplyInst *PAI) {
   if (foldInverseReabstractionThunks(PAI, this))
     return nullptr;
 
-  tryOptimizeApplyOfPartialApply(PAI, Builder, getInstModCallbacks());
+  SILBuilderContext BuilderCtxt(Builder.getModule(), Builder.getTrackingList());
```
Comment (gottesmm) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: You can get this directly from the builder.

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp (hunk 24 above)
Comment (gottesmm) [reply lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: Ok.

File: lib/SILOptimizer/Utils/PartialApplyCombiner.cpp (hunk 6 above)
Comment (gottesmm) [reply lib/SILOptimizer/Utils/PartialApplyCombiner.cpp]: I see. It was hoisted from below.

**Comment by vedantk** [comment]
Hi @eeckstein @gottesmm, I'm starting to see a use-before-init on the sanitizer bot after this PR:

```
/Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/swift/lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:122:9: runtime error: load of value 109, which is not a valid value for type 'bool'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/swift/lib/SILOptimizer/Mandatory/MandatoryCombine.cpp:122:9 in 
Stack dump:
0.	Program arguments: /Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/Ninja-ReleaseAssert+asan+ubsan/swift-macosx-x86_64/bin/swift -frontend -module-cache-path /Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/Ninja-ReleaseAssert+asan+ubsan/lldb-macosx-x86_64/./lldb-test-build.noindex/module-cache-clang -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -serialize-debugging-options -module-cache-path /Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/Ninja-ReleaseAssert+asan+ubsan/lldb-macosx-x86_64/test/Swift/Output/RemoteASTImport.test.tmp/cache -merge-modules -emit-module -parse-as-library -sil-merge-partial-modules -disable-diagnostic-passes -disable-sil-perf-optzns -module-name Library Library.part.swiftmodule -o Library.swiftmodule -I/Users/buildnode/jenkins/workspace/oss-lldb-asan-osx/Ninja-ReleaseAssert+asan+ubsan/lldb-macosx-x86_64/test/Swift/Output/RemoteASTImport.test.tmp 
```

This is here:

```
  bool runOnFunction(SILFunction &function) {
    bool changed = false;

    while (doOneIteration(function, iteration)) {
      changed = true;
      ++iteration;
    }

    if (invalidatedStackNesting) { //< here
      StackNesting().correctStackNesting(&function);
    }
```

**Comment by vedantk** [comment]
This also shows up when building the swift stdlib, so no lldb testing required to repro:

```
3.      While running pass #1 SILFunctionTransform "MandatoryCombine" on SILFunction "@$sSAyxGs8_PointerssABP9_rawValueBpvgTW".
 for getter for _rawValue (at /Users/vsk/src/github-swift-master/swift/stdlib/public/core/BridgeObjectiveC.swift:408:14)
```

**Comment by eeckstein** [comment]
@vedantk Thanks for letting me know.
https://github.com/apple/swift/pull/29842 should fix the issue
